src/components/MessagesList/MessagesList.vue

```diff
@@ -403,6 +403,22 @@ export default {
 		},
 
 		softUpdateAuthorGroups(oldGroups, newGroups, dateTimestamp) {
+			const newGroupIdSet = new Set(this.messagesList.map(message => message.id))
+			const newGroupsKeys = Object.keys(newGroups)
+
+			Object.entries(oldGroups).forEach(([id, oldGroup]) => {
+				if (newGroupsKeys.includes(id)) {
+					// Group is presented, will be handled in the next loop
+					return
+				}
+
+				// Check if group is older than any known newGroup
+				// or at least one message from group is no longer in the store
+				if (+id < +newGroupsKeys[0]
+					|| oldGroup.messages.some(message => !newGroupIdSet.has(message.id))) {
+					delete this.messagesGroupedByDateByAuthor[dateTimestamp][id]
+				}
+			})
```

Comment:
This is better to be done after updating with newGroups I think.
Also, we can minimize this loop by using `while` or `for` loop and once `+id > +newGroupsKeys[0]` break it so we just process the ones that really needs to be removed.

Reply:
This should happen before the soft update with newGroups, otherwise we might skip some overlapping groups, and messages will still be visually persisted.

Also see test cases above:
- we need to remove purged messages (that's always all messages before)
- but also expired messages (they can be in the middle of the list, so we might still know some recent messages)

Reply:
oh right, there is expired messages 🙈 
We can merge temp message removal with this purging loop:
```
			// Remove temporary messages that are not in the new list
			if (+dateTimestamp === this.currentDay) {
				const newGroupsMap = new Map(Object.entries(newGroups))
				for (const id of Object.keys(oldGroups).reverse()) {
					if (!id.toString().startsWith('temp-')) {
						break
					}
					if (!newGroupsMap.has(id)) {
						delete this.messagesGroupedByDateByAuthor[dateTimestamp][id]
					}
				}
			}
```